Create new Container Log Collection Troubleshooting Doc (#33283)
DataDog/documentation: content/en/containers/troubleshooting/log-collection.md

## Log collection context

In containerized environments, logs can be collected by the Datadog Agent in two main ways, file based collection and socket based collection through the Docker API.

The configuration steps provided in their respective Docker and Kubernetes docs default to file based collection as it offers better performance and reliability. Socket based collection can be used in Docker environments as a fallback option. In Kubernetes clusters socket based collection requires Docker as the container runtime, however this runtime is largely is deprecated in most Kubernetes distributions.

Datadog additionally recommends to have your applications log to the `stdout` / `stderr` streams in containerized environments. As opposed to writing to log files that are isolated in the application containers. Log collection with these streams can be setup much more automatically and reliably compared to the log files.

### Log files

> In Docker with the default `json-file` log driver, the `stdout`/`stderr` logs are stored in `/var/lib/docker/containers` and can be collected by mounting `/var/lib/docker/containers` (`c:/programdata/docker/containers` on Windows) into the Agent container. For example:

Reviewer: ```suggestion
With Docker's default `json-file` log driver, the `stdout`/`stderr` logs are stored in `/var/lib/docker/containers`. These logs can be collected by mounting `/var/lib/docker/containers` (`c:/programdata/docker/containers` on Windows) into the Agent container. For example:
```

```bash
/var/lib/docker/containers/68f21cd4e1e703c8b9ecdab67a5a9e359f1fb46ae1ed2e86f9b4f1f243a0a47d/68f21cd4e1e703c8b9ecdab67a5a9e359f1fb46ae1ed2e86f9b4f1f243a0a47d-json.log. 
```

> If this mount point does not exist, the Agent will fallback to socket based collection. Accessing the Docker API through the socket at `/var/run/docker.sock`. See [Docker Log Collection][1] for more information.

Reviewer: ```suggestion
If this mount point does not exist, the Agent falls back to socket based collection. Accessing the Docker API through the socket at `/var/run/docker.sock`.
```

> In Kubernetes, `stdout`/`stderr` logs are stored in `/var/log/pods` by default. The folder structure is setup per unique pod and container within it. For example:

Reviewer: ```suggestion
In Kubernetes, `stdout`/`stderr` logs are stored in `/var/log/pods` by default. The folder structure is setup per unique pod and container within that pod. For example:
```

```bash
/var/log/pods/default_my-deployment-55d847444b-2fkch_342819ce-4419-435b-9881-4a3deff618cc/my-container/0.log
```

If the container in the pod restarts in Kubernetes, it will automatically increment the filename (`0.log` -> `1.log`) which the Agent will account for automatically. See [Kubernetes log collection][2] for more information.

> As the Agent discovers the corresponding containers on the host, it will lookup their log files based on the expected folder and file structure per environment.

Reviewer: ```suggestion
As the Agent discovers the corresponding containers on the host, it looks up their log files based on the expected folder and file structure per environment.
```

### Agent autodiscovery

By default, the Agent only collects logs from containers when log collection is enabled in the Agent and either:

- The Datadog Agent has `logs_config.container_collect_all` enabled to collect logs from all discovered containers
- The container is configured for log collection from an Autodiscovery based integration

The agent also takes into account any container exclusion/inclusion rules you have configured from [Container Discovery Management][5]. 

Lastly, the Agent is responsible for collecting the logs from the containers on the same host as itself. 

> It is important to take these rules into account to understand how log collection is setup for your containers. If for example you do not see logs for a given container you should check:

Reviewer: I wasn't sure, are there more examples this applies to, should it be generic or is this specific to missing logs?

> It is important to take these rules into account to understand how log collection is setup for your containers. If for example you do not see logs for a given container you should check:

Author: This is generic. The example this is referring to is "if you do not see logs"

```

See more examples in [Docker Log Collection](/containers/docker/log/?tab=dockerfile#log-integrations).

{{% /tab %}}
{{< /tabs >}}

With both of these setups ensure that your configuration:
- Has at least the source and service set 
- Is valid JSON
- Is set on your corresponding Kubernetes pod or Docker container
- Uses the right key name to trigger the log configuration, you do not need to adjust the key name based on your [Datadog site][8]. 

Visit [Advanced Log Collection Configurations][9] for further examples of how to set your log configuration.

### Tagging

The Agent automatically assigns tags to your logs at the “high” level of [tag cardinality][10] for each environment. You can view the out-of-the-box [Docker tags here][11] and [Kubernetes tags here][12]. This additionally includes and tags collected by [Unified Service Tagging][13] or different tag extraction rules from container metadata.

To customize these tags, change the log collection rules, or enable log collection in general - you can apply Autodiscovery Labels or Annotations to the respective containers as documented above.

> Tags on your logs can also come from [host tag inheritance][14]. All data, including logs, coming into Datadog goes through this process. On Datadog intake the logs will inherit all the host-level tags that are associated with that host. You can see these tags on the Infrastructure List for you host. These are most commonly set by:

Reviewer: ```suggestion
To customize these tags, change the log collection rules, or enable log collection in general - you can apply Autodiscovery Labels or Annotations to the respective containers.

Tags on your logs can also come from [host tag inheritance][14]. All data, including logs, coming into Datadog goes through this process. On Datadog intake the logs inherit all the host-level tags that are associated with that host. You can see these tags on the Infrastructure List for you host. These are most commonly set by:
```

- The Datadog Agent and its automatic discovery or manual set of `DD_TAGS` provided
- The cloud provider integrations collecting and setting tags for your hosts

So for example the tags `pod_name` and `short_image` come from the Agent setting this tag on submission. Other tags like `region` and `kube_cluster_name` come from host tag inheritance on intake.

## Troubleshooting

The Datadog Agent running on the same node as your application container is the one responsible for collecting that container’s logs. So when running these commands, especially in Kubernetes environments, ensure you are working with the correct Agent pod for your desired application container.

> Visit [Agent Commands][15] to see a list of helpful agent commands for troubleshooting issues.

Reviewer: ```suggestion
For a list of helpful troubleshooting commands, see [Agent Commands][15].
```

```text
==========
Logs Agent
==========
    Reliable: Sending compressed logs in HTTPS to agent-http-intake.logs.datadoghq.com on port 443
    BytesSent: 8.60922316e+08
    EncodedBytesSent: 3.9744538e+07
    LogsProcessed: 604328
    LogsSent: 60431
  
  ============
  Integrations
  ============
  
  default/my-deployment-55d847444b-2fkch/my-container
  ---------------------------------------------------
    - Type: file
      Identifier: ba778eaff01fc3555b6ad4a809e78949065bd34ebe2c42522a1bdd1d3b684fb5
      Path: /var/log/pods/default_my-deployment-55d847444b-2fkch_342819ce-4419-435b-9881-4a3deff618cc/my-container/*.log
      Service: example-service
      Source: example-source
      Status: OK
        1 files tailed out of 1 files matching
      Inputs:
        /var/log/pods/default_my-deployment-55d847444b-2fkch_342819ce-4419-435b-9881-4a3deff618cc/my-container/0.log  
      Bytes Read: 5075   
      Pipeline Latency:
        Average Latency (ms): 0
        24h Average Latency (ms): 0
        Peak Latency (ms): 0
        24h Peak Latency (ms): 0
```

> If the Logs Agent Status doesn’t look like the above, refer to the troubleshooting tips in the following sections.

Reviewer: ```suggestion
If the Logs Agent Status doesn’t look like the above, see the troubleshooting tips in the following sections.
```

If you see the following message when you run the Agent status command:

```
==========
Logs Agent
==========

  Logs Agent is not running
```
This means that you did not enable log collection in the Agent.

If the Logs Agent Status shows no Integrations and you see `LogsProcessed: 0` and `LogsSent: 0`:
```
==========
Logs Agent
==========

    LogsProcessed: 0
    LogsSent: 0
```
This status means that logs are enabled but you haven’t specified which containers the Agent should collect from.

> Each individual log collector will provide you a lot of information. For example relative to that above sample from Kubernetes this is telling us:

Reviewer: Should this section go immediately after the example? Maybe before ln 241?

- The `default/my-deployment-55d847444b-2fkch/my-container`, is for the pod in the `default` namespace, pod named `my-deployment-55d847444b-2fkch`, with the container `my-container`. This is the individual log collector name.
- The `Identifier: ba778eaff01fc3555b6ad4a809e78949065bd34ebe2c42522a1bdd1d3b684fb5`, is the individual container ID that the log collection is for
- The `Path` and `Inputs` are the locations of where the Agent has searched for and identified the container logs

> - The `Service` and `Source` will tell you a quick view of the tags used

Reviewer: ```suggestion
- **Collector name** (`default/my-deployment-55d847444b-2fkch/my-container`) identifies namespace, pod, and container.
- **Identifier** (`ba778eaff...`) is the individual container ID being monitored.
- **Path** and **Inputs** show the locations where the Agent searched for and identified the container's log files.
- **Service** and **Source** summarize the tags used.
```

Author: Same as above. Will add a separate commit moving this part, and making your suggested changes.

In Docker the output is largely the same, just the individual log collector name is different.

### Agent configcheck

You can run the `agent configcheck` command to print all configurations loaded and resolved in a running Agent.

{{< tabs >}}
{{% tab "Kubernetes" %}}

```
kubectl exec -it <AGENT_POD> -- agent configcheck
```

{{% /tab %}}
{{% tab "Docker" %}}

```
docker exec -it <CONTAINER_NAME> agent configcheck
```

{{% /tab %}}
{{< /tabs >}}

This command shows you the configuration of the log collector, using the `Configuration source` referencing the container ID. This can be used to match up with the `agent status` output.

```
===  check ===
Configuration provider: kubernetes-container-allinone
Configuration source: container:containerd://ba778eaff01fc3555b6ad4a809e78949065bd34ebe2c42522a1bdd1d3b684fb5
Log Config:
[{"service":"example-service","source":"example-source"}]
Autodiscovery IDs:
* containerd://ba778eaff01fc3555b6ad4a809e78949065bd34ebe2c42522a1bdd1d3b684fb5
```

> You can see in this case, the `Log Config` applied from Autodiscovery provides custom `service` and `source` tags shown as `[{"service":"example-service","source":"example-source"}]`. The `configcheck` output is useful for verifying how the Agent setup log collection for a given container based on its container ID.

Reviewer: ```suggestion
The `Log Config` applied from Autodiscovery provides custom `service` and `source` tags shown as `[{"service":"example-service","source":"example-source"}]`. The `configcheck` output is useful for verifying how the Agent setup log collection for a given container based on its container ID.
```

When using `logs_config.container_collect_all` if no unique configuration is provided you will see this default to `[{}]` for the container.

### Agent stream-logs

You can run the `agent stream-logs` command to stream logs to the console that the Agent is seeing in real time, along with the associated metadata and the log content.

{{< tabs >}}
{{% tab "Kubernetes" %}}

```
kubectl exec -it <AGENT_POD> -- agent stream-logs

# Stream logs relative to "Namespace/Pod Name/Container Name" based name
kubectl exec -it <Agent Pod> -- agent stream-logs --name <NAME>
```

{{% /tab %}}
{{% tab "Docker" %}}

```
docker exec -it <CONTAINER_NAME> agent stream-logs
```

{{% /tab %}}
{{< /tabs >}}

You can filter this output using the `--name` flag to stream logs relative to the "Namespace/Pod Name/Container Name" based name in Kubernetes. You can alternatively use the `--service` or `--source` flag based on the tags used. 

> You can find the `<NAME>` using the `agent status` command. For example, in this case, the `<NAME>` is `default/my-deployment-55d847444b-2fkch/my-container`:

Reviewer: ```suggestion
To find the `<NAME>`, use the `agent status` command. For example, `default/my-deployment-55d847444b-2fkch/my-container`:
```

```

Add the Agent to the Docker user group, perform the following command:
```
usermod -a -G docker dd-agent
```
**Note:** When you install the Agent on the host, the Agent does not have permission to access` /var/lib/docker/containers` as these require root access. As a result it will collect logs from the Docker socket.

### Different Docker log driver

Docker’s default is the [json-file logging driver][23] so the Agent tries to read from this structure first. If your containers are set to use a different logging driver, the Logs Agent will indicate that it is able to successfully find your containers but it isn’t able to collect their logs from the file. In Docker environments, Datadog recommends using the `json-file` log driver for the optimal Agent experience. However, the Agent can also be configured to read from the `journald` logging driver.

1. If you’re unsure of which logging driver your containers are using, use `docker inspect <CONTAINER_NAME>` to see what logging driver you have set. The following block appears in the Docker Inspect when the container is using the JSON logging driver

```
"LogConfig": {
    "Type": "json-file",
    "Config": {}
},

> ```

Reviewer: This is to line it up under the numbered steps

````suggestion
   ```
   "LogConfig": {
       "Type": "json-file",
       "Config": {}
   },
   ```
````
